ruby/ql/lib/codeql/ruby/regexp/internal/RegExpTracking.qll

```diff
@@ -91,69 +55,175 @@ DataFlow::Node stringSink() {
 /** Gets a node where regular expressions that flow to the node are used. */
 DataFlow::Node regSink() { result = any(RegexExecution exec).getRegex() }
 
-/** Gets a node that is reachable by type-tracking from any string or regular expression. */
-DataFlow::LocalSourceNode forward(TypeTracker t) {
-  t.start() and
-  result = [strStart(), regStart()]
-  or
-  exists(TypeTracker t2 | result = forward(t2).track(t2, t))
-  or
-  exists(TypeTracker t2 | t2 = t.continue() | step(forward(t2).getALocalUse(), result, _, _))
+private signature module ReachInputSig {
+  DataFlow::LocalSourceNode start(TypeTracker t);
+
+  DataFlow::Node end();
+
+  predicate additionalStep(DataFlow::LocalSourceNode nodeFrom, DataFlow::LocalSourceNode nodeTo);
```

Comment:
The original `step` relation did not require a `LocalSourceNode` for `nodeFrom`. I am curious why you chose to ask for a big step here and then close under `flowsTo` on the user side?